### ISSUE TITLE
Support tensor parallel/pipeline parallel

### ISSUE THREAD

**Issue by gongel** [mosaicml/streaming#397]
Support tensor parallel/pipeline parallel currently?

**Comment by karan6181** [comment]
Can you please share more details ?

**Comment by gongel** [comment]
NVIDIA-Megatron team proposed "Tensor Parallelism". When training in "Tensor Parallelism", the rank in same group has same data.

**Comment by andreamad8** [comment]
Any plan to add this? 

one easy solution that seems not to be working could be:
```
os.environ["WORLD_SIZE"] = str(os.environ["WORLD_SIZE"]  // model_parallel_size)
os.environ["RANK"] = str(os.environ["RANK"] // model_parallel_size)
```

I tried, but seems the code gets stuck after calling something like:
```
batch = next(batch_iterator)
```

where batch_iterator a dataloder.

**Comment by karan6181** [comment]
@snarayan21 Looks like this is being addressed. Is that right?

**Comment by huxuan** [comment]
Would like to know if there is any example of megatron integration.

**Comment by snarayan21** [comment]
@andreamad8 @huxuan @gongel please see the `replication` argument detailed in our docs [here](https://docs.mosaicml.com/projects/streaming/en/stable/dataset_configuration/replication_and_sampling.html#replication).

@huxuan We don't have an explicit example of a megatron integration, but as it's pytorch based, you can simply swap in the dataset / dataloader.

**Comment by casper-hansen** [comment]
@snarayan21 

We can compute replication based on TP and SP as follows. But it seems we are lacking documentation on PP. I think part of this issue was unanswered because of that. How do we modify the parameters to streaming with pipeline parallelism?

TP and SP:

```
replication = tp * sp
```

Various configurations I tried just to sanity check. None of these are correct (unexpected high loss).

1. 

```
replication = tp * sp * pp
```

2. 

```
replication = tp * sp * pp
num_canonical_nodes = world_size // replication
```

3. 

```
replication = tp * sp
num_canonical_nodes = world_size // (tp * sp * pp)
```

**Comment by casper-hansen** [comment]
Opened an issue for this separately:
https://github.com/mosaicml/streaming/issues/827
